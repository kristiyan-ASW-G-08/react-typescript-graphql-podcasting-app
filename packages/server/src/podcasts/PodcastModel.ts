import mongoose, { Schema } from 'mongoose';
import PodcastType from '@customTypes/PodcastType';
import uniqueValidator from 'mongoose-unique-validator';
import duplicationErrorHandler from '@customUtilities/duplicationErrorHandler';

const PodcastSchema = new Schema<PodcastType>({
  title: { type: String, maxlength: 50, minlength: 1, unique: true },
  user: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  website: { type: String, required: true },
  cover: { type: String, required: true },
});

// @ts-ignore
PodcastSchema.plugin(uniqueValidator);
// @ts-ignore
PodcastSchema.post('save', duplicationErrorHandler);
// @ts-ignore
PodcastSchema.post('update', duplicationErrorHandler);

PodcastSchema.index({ title: 'text' });

export default mongoose.model<PodcastType>('Podcast', PodcastSchema);
