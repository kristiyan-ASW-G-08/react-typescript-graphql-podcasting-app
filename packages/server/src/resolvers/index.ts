import userMutations from 'src/users/usersMutations';
import podcastMutations from '@src/podcasts/podcastsMutations';

const resolvers = {
  Query: {
    getUser: async () => {},
  },
  Mutation: { ...userMutations, ...podcastMutations },
};

export default resolvers;
