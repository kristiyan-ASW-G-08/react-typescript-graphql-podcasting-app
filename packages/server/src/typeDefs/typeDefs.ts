import { gql } from 'apollo-server';

const typeDefs = gql`
  scalar Upload

  type User {
    username: String!
    email: String!
    id: ID!
  }

  type Query {
    getUser: User!
  }

  type File {
    filename: String!
    mimetype: String!
    encoding: String!
  }
  type UserLoginPayload {
    username: String!
    email: String!
    token: String!
  }
  type Podcast {
    title: String!
    website: String!
    user: User!
    _id: ID!
  }
  type Query {
    podcasts(query: String): [Podcast]!
  }
  type Mutation {
    createUserMutation(
      username: String!
      email: String!
      password: String!
      confirmPassword: String!
    ): UserLoginPayload!

    loginMutation(email: String!, password: String!): UserLoginPayload!

    createPodcastMutation(
      title: String!
      website: String!
      cover: Upload!
    ): Podcast!

    updatePodcastMutation(
      title: String!
      website: String!
      cover: Upload
      _id: ID!
    ): Podcast!
  }
`;

export default typeDefs;
