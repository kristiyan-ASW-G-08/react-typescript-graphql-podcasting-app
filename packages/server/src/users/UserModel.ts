import mongoose, { Schema } from 'mongoose';
import UserType from '@src/types/UserType';
import duplicationErrorHandler from '@customUtilities/duplicationErrorHandler';
import uniqueValidator from 'mongoose-unique-validator';

const UserSchema: Schema = new Schema({
  username: { type: String, minlength: 1, maxlength: 50, unique: true },
  email: { type: String, unique: true },
  password: { type: String, minlength: 8 },
});

UserSchema.plugin(uniqueValidator);
// @ts-ignore
UserSchema.post('save', duplicationErrorHandler);
// @ts-ignore
UserSchema.post('update', duplicationErrorHandler);

export default mongoose.model<UserType>('User', UserSchema);
