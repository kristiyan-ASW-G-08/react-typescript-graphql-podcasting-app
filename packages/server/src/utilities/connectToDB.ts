import mongoose from 'mongoose';
import logger from '@customUtilities/logger';

const connectToDB = async (mongoURI: string): Promise<void> => {
  try {
    await mongoose.connect(mongoURI, {
      useNewUrlParser: true,
      useUnifiedTopology: true,
    });
  } catch (err) {
    logger.error(`MongoDB Connection Error: ${err}`);
  }
};
export default connectToDB;
