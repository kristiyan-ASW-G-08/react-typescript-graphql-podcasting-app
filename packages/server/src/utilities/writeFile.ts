import fs from 'fs';
import path from 'path';

const writeFile = (
  filename: string,
  stream: NodeJS.ReadWriteStream,
): string => {
  // @ts-ignore
  const filePath = path.join(
    path.basename(path.dirname(filename)),
    `/images/${new Date().toISOString()}-${filename}`,
  );
  stream.pipe(fs.createWriteStream(filePath));
  return filePath;
};

export default writeFile;
