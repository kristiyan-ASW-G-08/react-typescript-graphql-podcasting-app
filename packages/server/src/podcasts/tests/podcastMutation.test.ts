import mongoose from 'mongoose';
import { ApolloServer } from 'apollo-server-express';
import { sign } from 'jsonwebtoken';
import FormData from 'form-data';
import connectToDB from '@customUtilities/connectToDB';
import UserModel from '@src/users/UserModel';
import PodcastModel from '@src/podcasts/PodcastModel';
// @ts-ignore
import typeDefs from '@typeDefs/typeDefs';
import apolloServer from '@customUtilities/createApoloServer';
import { string } from 'yup';
import fs from 'fs';
import { Readable } from 'stream';
// @ts-ignore
import { Upload } from 'graphql-upload/public';
import writeFile from '@customUtilities/writeFile';
import resolvers from '../../resolvers/index';

jest.mock('@customUtilities/writeFile');

(writeFile as jest.Mock).mockReturnValue('/images/some_file.svg');
describe('Resolvers', () => {
  const { MONGO_USER, MONGO_PASSWORD, MONGO_DATABASE } = process.env;
  const mongoURI = `mongodb+srv://${MONGO_USER}:${MONGO_PASSWORD}@cluster0.ol9wi.mongodb.net/${MONGO_DATABASE}?retryWrites=true`;
  beforeAll(async () => {
    await mongoose.disconnect();
    await connectToDB(mongoURI);
    await UserModel.deleteMany({}).exec();
  });

  afterEach(async () => {
    await UserModel.deleteMany({}).exec();
  });
  afterAll(async () => {
    await mongoose.disconnect();
  });

  describe('Mutations', () => {
    const { SECRET } = process.env;
    const token = sign(
      {
        userId: mongoose.Types.ObjectId(),
      },
      SECRET,
      { expiresIn: '96h' },
    );

    afterEach(async () => {
      await PodcastModel.deleteMany({}).exec();
    });

    const values = {
      title: 'someTitle',
      website: 'http://somefakewebsite.fakewebsite',
    };

    const incorrectValues = {
      username: '',
      email: 'notAnEMail',
      password: 'pass',
      confirmPAssword: '',
    };
    describe('createPodcastMutation', () => {
      it('create podcast successfully', async () => {
        expect.assertions(2);
        const file = Readable.from(Buffer.from('hello upload', 'utf-8'));
        // @ts-ignore

        const upload = new Upload();
        upload.promise = new Promise(resolve =>
          resolve({
            createReadStream: () => file,
            filename: 'some_file.svg',
            mimetype: 'text/plain',
          }),
        );
        const query = `
          mutation createPodcastMutation( $title:String!,$website:String!,$cover: Upload!) {
            createPodcastMutation(title: $title,website: $website,cover: $cover) {
              _id
            }
          }
        `;

        const testServer = new ApolloServer({
          typeDefs,
          resolvers,
          context: () => ({ headers: { authorization: `bearer ${token}` } }),
        });

        const result = await testServer.executeOperation({
          query,
          variables: {
            cover: upload,
            ...values,
          },
        });
        console.log(JSON.stringify(result.errors));
        expect(result.errors).toBeUndefined();
        expect(result.data?.createPodcastMutation._id).toBeTruthy();
      });
    });

    it('fail to create a podcast due to validation errors', async () => {
      expect.assertions(1);

      const query = `
        mutation createPodcastMutation( $title:String!,$website:String!,$cover: Upload!) {
          createPodcastMutation(title: $title,website: $website,cover: $cover) {
            _id
          }
        }
      `;

      const testServer = new ApolloServer({
        typeDefs,
        resolvers,
        context: () => ({ headers: { authorization: `bearer ${token}` } }),
      });

      const result = await testServer.executeOperation({
        query,
        variables: {},
      });
      expect(result.errors).toMatchSnapshot();
    });
    it('fail to create a podcast due to authorization error', async () => {
      expect.assertions(1);
      const file = Readable.from(Buffer.from('hello upload', 'utf-8'));
      // @ts-ignore

      const upload = new Upload();
      upload.promise = new Promise(resolve =>
        resolve({
          createReadStream: () => file,
          filename: 'some_file.svg',
          mimetype: 'text/plain',
        }),
      );
      const query = `
        mutation createPodcastMutation( $title:String!,$website:String!,$cover: Upload!) {
          createPodcastMutation(title: $title,website: $website,cover: $cover) {
            _id
          }
        }
      `;

      const testServer = new ApolloServer({
        typeDefs,
        resolvers,
      });

      const result = await testServer.executeOperation({
        query,
        variables: {
          cover: upload,
          ...values,
        },
      });
      expect(result.errors).toMatchSnapshot();
    });
  });
});
