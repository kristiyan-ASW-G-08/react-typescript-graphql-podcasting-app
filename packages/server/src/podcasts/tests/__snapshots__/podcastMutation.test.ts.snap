// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolvers Mutations fail to create a podcast due to authorization  1`] = `
Array [
  [GraphQLError: Cannot read properties of undefined (reading 'authorization')],
]
`;

exports[`Resolvers Mutations fail to create a podcast due to validation errors 1`] = `
Array [
  [UserInputError: Variable "$title" of required type "String!" was not provided.],
  [UserInputError: Variable "$website" of required type "String!" was not provided.],
  [UserInputError: Variable "$cover" of required type "Upload!" was not provided.],
]
`;
